tasks.py

```diff
@@ -154,24 +156,111 @@ def _get_latest_version() -> Version:
+    r = requests.get("https://pypi.org/pypi/great-expectations-cloud/json", timeout=10)
+    r.raise_for_status()
+    return [Version(v) for v in r.json()["releases"].keys()]
+
+
+def _get_latest_versions() -> tuple[Version, Version]:
+    all_versions = _get_all_versions()
+    pre_releases = sorted(v for v in all_versions if v.is_prerelease)
+    releases = sorted(v for v in all_versions if not v.is_prerelease)
+    return max(pre_releases), max(releases)
+
+
+def _new_release_version(
+    latest_version: Version,
+    current_date: str,
+) -> Version:
+    proposed_version = Version(current_date)
+    if proposed_version > latest_version:
+        # Always bump if it is a new date
+        return Version(f"{proposed_version.major}.0")
+    else:
+        # Bump the minor version only if it is the same date
+        return Version(f"{latest_version.major}.{latest_version.minor + 1}")
+
+
+class PreReleaseVersionError(Exception):
+    def __init__(
+        self, latest_version: Version, latest_pre_release_version: Version, current_date: str
+    ):
+        super().__init__("Could not determine the new pre-release version.")
+        self.latest_version = latest_version
+        self.latest_pre_release_version = latest_pre_release_version
+        self.current_date = current_date
+
+    @override
+    def __str__(self) -> str:
+        return f"{super().__str__()} latest_version: {self.latest_version} latest_pre_release_version: {self.latest_pre_release_version} current_date: {self.current_date}"
+
+
+def _new_pre_release_version(
+    latest_version: Version,
+    latest_pre_release_version: Version,
+    current_date: str,
+) -> Version:
+    # Case where the latest pre-release is ahead of the latest version e.g. the next day
+    if latest_pre_release_version > latest_version:
+        current_dev_version = latest_pre_release_version.dev or 0
+        return Version(
+            f"{latest_pre_release_version.major}.{latest_pre_release_version.minor}.dev{current_dev_version + 1}"
+        )
+    # Case where the latest pre-release was on the same date and there was already a release on that date
+    elif latest_version.major == int(current_date) and latest_version.minor:
+        new_version = Version(f"{current_date}.{latest_version.minor + 1}.dev0")
+        return new_version
+    # Case where the latest pre-release was on the same date
+    elif latest_version.major == int(current_date):
+        new_version = Version(f"{current_date}.1.dev0")
+        return new_version
+    # Case where the latest pre-release was on an earlier date
+    elif latest_pre_release_version.major < int(current_date):
+        new_version = Version(f"{current_date}.0.dev0")
+        return new_version
+    else:
+        # We should never get here
+        raise PreReleaseVersionError(
+            latest_version=latest_version,
+            latest_pre_release_version=latest_pre_release_version,
+            current_date=current_date,
+        )
```

Comment:
I simplified this logic quite a bit, but happy to simplify further if there is any low hanging fruit.

Reply:
My only suggestion would be spending a bit of time looking into existing python packages that can do this thing. It feels error prone. We might still decide this is fine as it is.
Like what does `Versions.__add__` do?

Or something like `bump-my-version` (formerly `bump-version`) should work, but you've already done a bunch of the work.
https://callowayproject.github.io/bump-my-version/

Reply:
I'll also review your comments more thoroughly but I did look into [bumpver](https://github.com/mbarkhau/bumpver) and a few others that turned out to be not very well maintained. It seemed like it would be a little complicated to do what we are planning here in bumpver so I just implemented. I can investigate bump-my-version, but figure this is pretty simple and worst case we can bump manually if any of the logic is off when you run the tool so figure it is relatively low priority to migrate.

Reply:
Thanks for the input here @TrangPham @Kilo59 - I simplified the logic to only use the latest release. Let me know what you think!
I think for now if you are good with the logic I'd like to release this just to keep moving forward and investigate more deeply into replacing with bump-my-version after we get a bit more of the epic work done.

Reply:
(Mainly because I'm not seeing an obvious way right off the bat to incorporate pre-releases with calver in bump-my-version: https://callowayproject.github.io/bump-my-version/howtos/calver/)